hail/python/hailtop/aiotools/fs.py

```diff
@@ -130,13 +130,30 @@ async def remove(self, url: str) -> None:
         pass
 
     @abc.abstractmethod
-    async def rmtree(self, sema: asyncio.Semaphore, url: str) -> None:
+    async def rmtree(self, sema: Optional[asyncio.Semaphore], url: str) -> None:
         pass
 
     async def touch(self, url: str) -> None:
         async with await self.create(url):
             pass
 
+    async def read(self, url: str) -> bytes:
+        async with await self.open(url) as f:
```

Comment:
I think I agree with @danking.  This implementation should work in the RouterFS because the call to `self.open` will dispatch off the URL scheme.